src/Elastic.Apm/DistributedTracing/TraceParent.cs

```diff
@@ -0,0 +1,199 @@
+	/// "https://www.w3.org/TR/trace-context/#traceparent-field" w3c 'traceparent' header draft.
+	/// elastic-apm-traceparent: 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01
+	/// (______________________)  () (______________________________) (______________) ()
+	///      v                     v                 v                        v         v
+	///  Header name           Version           Trace-Id                Span-Id     Flags
+	/// Since the w3c document is just a draft at the moment,
+	/// we don't use the official header name but prepend the custom prefix "Elastic-Apm-".
+	/// </summary>
+	internal static class TraceParent
+	{
+		private const int VersionPrefixIdLength = 3;
+		private const int VersionLength = 2;
+		private const int VersionAndTraceIdLength = 36;
+
+		private const int TraceIdLength = 32;
+		private const int VersionAndTraceIdAndSpanIdLength = 53;
+		private const int SpanIdLength = 16;
+		private const int OptionsLength = 2;
+
+		private const byte FlagRecorded = 1; // 00000001
+		internal const string TraceParentHeaderName = "elastic-apm-traceparent";
+
+		/// <summary>
+		/// Checks if the <paramref name="traceOptions" /> flag contains the <see cref="FlagRecorded" /> flag.
+		/// </summary>
+		/// <param name="traceOptions">The traceOptions flags return from <see cref="TryExtractTraceparent" />.</param>
+		/// <returns>true if <paramref name="traceOptions" /> contains <see cref="FlagRecorded" />, false otherwise.</returns>
+		public static bool IsFlagRecordedActive(byte[] traceOptions) => (traceOptions[0] & FlagRecorded) == FlagRecorded;
+
+		/// <summary>
+		/// Parses the traceparent header
+		/// </summary>
+		/// <param name="traceParentValue">The value of the traceparent header</param>
+		/// <param name="traceId">The parsed traceId</param>
+		/// <param name="prentId">The parsed parentId</param>
+		/// <param name="traceOptions">The parsed traceOptions flags</param>
+		/// <returns>True if parsing was successful, false otherwise.</returns>
+		internal static bool TryExtractTraceparent(string traceParentValue, out string traceId, out string prentId, out byte[] traceOptions)
+		{
+			traceId = string.Empty;
+			prentId = string.Empty;
+			traceOptions = null;
+
+			var bestAttempt = false;
+
+			if (string.IsNullOrWhiteSpace(traceParentValue)) return false;
+
+			if (traceParentValue.Length < VersionPrefixIdLength || traceParentValue[VersionPrefixIdLength - 1] != '-') return false;
+
+			try
+			{
+				var versionArray = StringToByteArray(traceParentValue, 0, VersionLength);
+				if (versionArray[0] == 255)
+					return false;
+
+				if (versionArray[0] > 0)
+					// expected version is 00
+					// for higher versions - best attempt parsing of trace id, span id, etc.
+					bestAttempt = true;
+			}
+			catch
+			{
+				return false;
+			}
+
+			if (traceParentValue.Length < VersionAndTraceIdLength || traceParentValue[VersionAndTraceIdLength - 1] != '-') return false;
```

Comment:
That is mainly inspired by the OpenCensus code 🤣  https://github.com/census-instrumentation/opencensus-csharp/blob/4a8ddf6727eafda97a06c7c30d8a4fc2ec8b8e2f/src/OpenCensus/Trace/Propagation/TraceContextFormat.cs 

I think this is nice, because we access `traceParentValue[VersionAndTraceIdLength - 1]`, so before that we make sure that the string has that id. I see your point, it could just exit at the top, that'd be the same. Honestly I'd prefer this logic.